enhancements/ingress/lb-allowed-source-ranges.md

```diff
@@ -0,0 +1,227 @@
+  - "@deads2k"
+creation-date: 2022-07-08
+last-updated: 2022-07-26
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - "https://issues.redhat.com/browse/NE-555"
+see-also:
+replaces:
+superseded-by:
+---
+
+# LoadBalancer Allowed Source Ranges
+
+## Summary
+
+This enhancement extends the IngressController API to allow the user to specify a 
+list of IP address ranges for an IngressController to which access to the LoadBalancer Service 
+should be restricted when `endpointPublishingStrategy` is `LoadBalancerService`. 
+By default, all source addresses are allowed (which is equivalent to
+specifying "0.0.0.0/0"). Currently, the only way for the customers to limit this is by 
+setting `.Spec.LoadBalancerSourceRanges` field for the LoadBalancer Service 
+in front of the router in cloud environments. This is not desired as we prefer
+the customers to use the APIs and to not directly modify the managed resources.
+
+## Motivation
+
+Customer security teams are concerned that the public ELB (created in 
+the installation process by default) is allowing all inbound traffic on 0.0.0.0 
+on ports 80 and 443, so they wish to restrict it. 
+
+Customers have been setting `.Spec.LoadBalancerSourceRanges` of the LoadBalancer 
+Service themselves, something that we did not plan to support but also did not 
+actively reject. We recognize that users consider it now to be an existing feature 
+while we prefer they use APIs and not directly modify managed resources.
+
+The main motivation of this enhancement is to allow the customers to enable 
+and configure `.Spec.LoadBalancerSourceRanges` through a new API.
+
+### User Stories
+
+#### As a cluster admin, when configuring an IngressController, I want to limit access to the load balancer to a specified list of IP address ranges on the cloud provider for this ingress
+
+The cluster admin has the option to modify the IngressController to specify 
+a list of IP address ranges for an IngressController to which access to 
+the load balancer should be restricted. This is available as a Day-2 operation 
+to the cluster admin.
+
+### Goals
+
+Enable the configuration of a range for a load balancer via the `IngressController.Spec`, 
+specifically via the `LoadBalancerStrategy`, with a new parameter `AllowedSourceRanges`, 
+which represents a list of IP address ranges. This is then used by the operator to set
+`.Spec.LoadBalancerSourceRanges` of the LoadBalancer Service in front of the router 
+in cloud environments.
+
+If `.Spec.LoadBalancerSourceRanges` of the LoadBalancer Service is already set by the 
+customer, its value is used to set `AllowedSourceRanges` if and only if `AllowedSourceRanges`
+was not initially set. Otherwise, `.Spec.LoadBalancerSourceRanges` is overwritten by `AllowedSourceRanges`.
```

Comment:
In my opinion, this is too complicated and error-prone.  Here is what I had in mind:

* On OpenShift 4.11, block upgrade if `spec.loadBalancerSourceRanges` is set.
* On OpenShift 4.12, block upgrade if `service.beta.kubernetes.io/load-balancer-source-ranges` is set.
* Don't set `spec.loadBalancerSourceRanges` if `service.beta.kubernetes.io/load-balancer-source-ranges` is set.

Then the cluster administrator can migrate to the new API as follows:

1. Make sure that `service.beta.kubernetes.io/load-balancer-source-ranges` is set and `spec.loadBalancerSourceRanges` is unset.
2. Upgrade to 4.12.
3. Set `spec.endpointPublishingStrategy.loadBalancer.allowedSourceRanges` on the IngressController.  (The ingress operator will then set `spec.loadBalancerSourceRanges`, which takes precedence over the annotation.)  
4. Remove the `service.beta.kubernetes.io/load-balancer-source-ranges` service annotation.  

These are all implementation details though.  In my mind the goals are the following:

* Provide a supported API for configuring allowed source ranges for IngressControllers
* Migrate customers who currently are configuring allowed source ranges on the Service to using this new API.

Reply:
In the 3rd step, you say 

> (The ingress operator will then set spec.loadBalancerSourceRanges, which takes precedence over the annotation.)

At this point, the annotation is still set though. So, the ingress operator should not set `spec.loadBalancerSourceRanges` according to the other statement 

> Don't set spec.loadBalancerSourceRanges if service.beta.kubernetes.io/load-balancer-source-ranges is set.

I assume the operator should actually set `spec.loadBalancerSourceRanges` only after the 4th step when annotation gets removed. 

Do I understand it right or did you think it differently?

Reply:
On further thought, maybe we can strike my third bulletpoint about not setting `spec.loadBalancerSourceRanges` if `service.beta.kubernetes.io/load-balancer-source-ranges` is set.  I believe the third bulletpoint is unnecessary because `spec.endpointPublishingStrategy.loadBalancer.allowedSourceRanges` and `spec.loadBalancerSourceRanges` would be empty after the 4.11→4.12 upgrade, and in that case the annotation takes precedence; and the bulletpoint may be problematic because it may cause a race condition if the user deletes the annotation and then the service controller observes that update before the ingress operator is able to update `spec.loadBalancerSourceRanges`.